pennylane/compiler/qjit_api.py

```diff
@@ -111,3 +112,173 @@ def circuit(theta):
+
+    Args:
+        cond_fn (Callable): the condition function in the while loop
+
+    Returns:
+        Callable: A wrapper around the while-loop function.
+
+    Raises:
+        CompileError: if the compiler is not installed
+
+    .. seealso:: :func:`~.for_loop`, :func:`~.qjit`
+
+    **Example**
+
+    .. code-block:: python
+
+        dev = qml.device("lightning.qubit", wires=1)
+
+        @qml.qjit
+        @qml.qnode(dev)
+        def circuit(x: float):
+
+            @qml.while_loop(lambda x: x < 2.0)
+            def loop_rx(x):
+                # perform some work and update (some of) the arguments
+                qml.RX(x, wires=0)
+                return x ** 2
+
+            # apply the while loop
+            final_x = loop_rx(x)
+
+            return qml.expval(qml.PauliZ(0)), final_x
+
+    >>> circuit(1.6)
+    [array(-0.02919952), array(2.56)]
+    """
+
+    if active_jit := active_compiler():
+        compilers = AvailableCompilers.names_entrypoints
+        ops_loader = compilers[active_jit]["ops"].load()
+        return ops_loader.while_loop(*args, **kwargs)
+
+    raise CompileError("There is no active compiler package.")
+
+
+def for_loop(*args, **kwargs):
+    """A :func:`~.qjit` compatible for-loop for PennyLane programs.
+
+    .. note::
+
+        This function only supports the Catalyst compiler. See
+        :func:`catalyst.for_loop` for more details.
+
+        Please see the Catalyst :doc:`quickstart guide <catalyst:dev/quick_start>`,
+        as well as the :doc:`sharp bits and debugging tips <catalyst:dev/sharp_bits>`
+        page for an overview of the differences between Catalyst and PennyLane.
+
+    This decorator provides a functional version of the traditional
+    for-loop, similar to ``jax.cond.fori_loop``. That is, any variables that
+    are modified across iterations need to be provided as inputs/outputs to
+    the loop body function:
+
+    - Input arguments contain the value of a variable at the start of an
+      iteration.
+
+    - output arguments contain the value at the end of the iteration. The
+      outputs are then fed back as inputs to the next iteration.
+
+    The final iteration values are also returned from the transformed
+    function.
+
+    The semantics of ``for_loop`` are given by the following Python pseudo-code:
+
+    .. code-block:: python
+
+        def for_loop(lower_bound, upper_bound, step, loop_fn, *args):
+            for i in range(lower_bound, upper_bound, step):
+                args = loop_fn(i, *args)
+            return args
+
+    Unlike ``jax.cond.fori_loop``, the step can be negative if it is known at tracing time
+    (i.e., constant). If a non-constant negative step is used, the loop will produce no iterations.
+
+    Args:
+        lower_bound (int): starting value of the iteration index
+        upper_bound (int): (exclusive) upper bound of the iteration index
+        step (int): increment applied to the iteration index at the end of each iteration
```

Comment:
Do we need `loop_fn`?

Reply:
I copied from the Catalyst docs. Happy to update the code after resolving this: https://github.com/PennyLaneAI/pennylane/pull/4698/files#r1393090074

Reply:
Fixed [c2df222](https://github.com/PennyLaneAI/pennylane/pull/4698/commits/c2df22226b69f4397f1546e4d4115344875c834a)